Lore Section of Bestiary

Wembler23 opened the pull request:
This adds a lore section to the bestiary, accessible through /bestiary lore <monster name> <lore>. The newline character is `~~~` because \n wouldn't work. The lore section is in the form of a book item in the inventory where the lore of that item is the lore submitted.

Combustible committed:
Add LibraryOfSoulsAPI class

Signed-off-by: Byron Marohn <combustible@live.com>

Wembler23 committed:
Lore Section of Bestiary

This adds a lore section to the bestiary, accessible through /bestiary lore <monster name> <lore>. The newline character is `~~~` because \n wouldn't work. The lore section is in the form of a book item in the inventory where the lore of that item is the lore submitted.

Wembler23 committed:
Merge branch 'master' into bestiary-lore-implementation

Wembler23 committed:
Merge branch 'master' into bestiary-lore-implementation

Wembler23 committed:
Merge remote-tracking branch 'origin/bestiary-lore-implementation' into bestiary-lore-implementation

Wembler23 committed:
Merge fixes

god i love merging

Wembler23 committed:
Checkstyle and bug fixes

Fixed checkstyle and also replaced a thrown error with a placeholder string - i think this is fine?

Wembler23 committed:
More checkstyle fixes

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/LibraryOfSoulsAPI.java:
```diff
@@ -24,14 +24,15 @@ public static Entity summon(Location loc, String soulName) {
 	}
 
-	public static Set<String> getSoulNames() {
+	public Set<String> getSoulNames() throws IllegalStateException {
```
Why was this change made? getSoulNames() has to be static - this is the entire point of this API.

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/Soul.java:
```diff
@@ -31,5 +31,9 @@
+	/* This is the label-ified name, with colors and spaces stripped */
+	@Override
+	String getLabel();
+
```
I don't understand why this is here - Soul already extends SoulGroup so it already has getLabel()

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/bestiary/BestiaryCommand.java:
```diff
@@ -89,6 +91,26 @@ public static void register() {
+					.executes((sender, args) -> {
+						if (!(sender instanceof Player)) {
+							sender.sendMessage("Gotta do this as a player until I remember how to make this work in command blocks. If you need that much for lore tell me.");
+						}
```
instead of executes() use executesPlayer() which gives you a player directly, instead of a sender.

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/bestiary/BestiaryCommand.java:
```diff
@@ -89,6 +91,26 @@ public static void register() {
+				.withSubcommand(new CommandAPICommand("clear")
+				.withPermission(CommandPermission.fromString("los.bestiary.lore"))
+				.withArguments(new StringArgument("mobLabel").replaceSuggestions(LibraryOfSoulsCommand.LIST_MOBS_FUNCTION))
+				.executes((sender, args) -> {
```
executesPlayer

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/SoulEntry.java:
```diff
@@ -248,6 +265,7 @@ public void openBestiary(Player player, BestiaryArea parent, List<BestiaryEntryI
 	 *--------------------------------------------------------------------------------*/
 
 	public enum InfoTier {
+		LORE(4),
```
I don't think we should add another tier here - would rather "Everything" includes the lore too.

Combustible replied:
Looks like this is still here?

Wembler23 committed:
Fixed requests, and added an invulnerable mobs override

Removed the lore tier and did some cleanup. Also made it so that if a mob is invulnerable, you are able to see the full entry as soon as you unlock the area, and it gets counted to your total found metric as well.

Combustible commented on plugin/src/main/java/com/playmonumenta/libraryofsouls/bestiary/BestiaryAreaInventory.java:
```diff
@@ -177,11 +177,12 @@ public int accountForOffset(int slot) {
 	}
 
-	public double totalDiscovered(BestiaryEntryInterface entry, Player player) {
+	//Workaround because it needs what the previous entry was to smash a bug
+	public double totalDiscovered(BestiaryEntryInterface entry, Player player, BestiaryEntryInterface lastChecked) {
```
Is this still true? Looks like lastChecked isn't used.

Wembler23 committed:
Fixed requests

Removed and edited the things requested, and hid the lore item if there is no lore to be found for that mob.
